mini-projects/00-jq-rs/src/filter.rs

```diff
@@ -0,0 +1,84 @@
+use std::str::Chars;
+
+use serde_json::Value;
+
+pub fn filter(json: &Value, filter: &str) -> Value {
+    let mut values = json.clone();
+    let filter_parts = filter.split('|');
+    for next_filter in filter_parts {
+        values = apply_filter(&values, next_filter.trim());
+    }
+    values
+}
+
+fn apply_filter(json: &Value, filter: &str) -> Value {
+    let mut values = json.clone();
+    let mut filter_chars = filter.chars();
+    if filter_chars.next() == Some('.') {
+        if filter_chars.clone().count() == 0 {
+            // Return the entire JSON if the filter is empty
+        } else if filter_chars.next() == Some('[') {
+            values = index_array(json, filter_chars);
+        } else {
+            let filter = &filter.chars().skip(1).collect::<String>();
+            if values.is_object() {
+                values = values.get(filter).unwrap_or(&Value::Null).clone();
+            } else {
+                panic!("Cannot index into a non-object");
+            }
+        }
```

Comment:
as a whole, this a great first pass at being able to parse multiple different filter types, but this logic could definitely have been made more explicit (maybe through a regex pattern search). I can imagine that, with the way it's currently architected, this wouldn't scale well to 10-20 different filter options. Something that could improve this is having an explicit enum to represent all the filters your application supports, and then being able to explicitly check if the enum's conditions for the filter are met